### PR TITLE
Fix 3609 - legendLayerZoomNative zooms to correct scale

### PR DESCRIPTION
**Pull request by stevenmizuno**
This uses the pixel diagonal distance for the source raster to match the diagonal of the mapCanvas  with OTF

### CODE REVIEW

**Comment by NathanW2** [comment]
Context: http://hub.qgis.org/issues/3609#note-12

**Comment by NathanW2** [comment]
@stevenmizuno if you are a first time contributor to the project you can label your PR with **New Contributor** label

**Comment by stevenmizuno** [comment]
This is my first time using GitHub, although I have contributed several things to the project some time ago. 

I was looking for information on procedures to submit pull requests (beyond the GitHub help) and found only a brief mention of using GitHub in the Developing C++ plugins documentation. Maybe I'm looking in the wrong place.

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -7643,9 +7643,8 @@ void QgisApp::legendLayerZoomNative()
       QgsCoordinateTransform ct( mMapCanvas->mapSettings().destinationCrs(), layer->crs() );
       p1 = ct.transform( p1 );
       p2 = ct.transform( p2 );
-      double width = sqrt( p1.sqrDist( p2 ) ); // width of reprojected pixel
-      // This is not perfect of course, we use the resolution in just one direction
-      mMapCanvas->zoomByFactor( qAbs( layer->rasterUnitsPerPixelX() / width ) );
+      double width = sqrt( p1.sqrDist( p2 ) ); // width (actually the diagonal) of reprojected pixel
```
Comment (vmora) [inline src/app/qgisapp.cpp]: What about: 
    const double diagonalLengthOfReprojectedPixel = sqrt( p1.sqrDist( p2 ) );
and leaving out the comment ?

File: src/app/qgisapp.cpp [hunk 2]
```diff
@@ -7643,9 +7643,8 @@ void QgisApp::legendLayerZoomNative()
...
+      mMapCanvas->zoomByFactor( sqrt( layer->rasterUnitsPerPixelX() * layer->rasterUnitsPerPixelX() + layer->rasterUnitsPerPixelY() * layer->rasterUnitsPerPixelY() ) / width );
```
Comment (vmora) [inline src/app/qgisapp.cpp]: mMapCanvas->zoomByFactor( QVector2D(layer->rasterUnitsPerPixelX(), layer->rasterUnitsPerPixelX()).length() 
                            / diagonalLengthOfReprojectedPixel );
Should work too, and compensates for the longer variable name.

**Comment by vmora** [comment]
@stevenmizuno I believe the merges are unnecessary.

**Comment by stevenmizuno** [comment]
@vmora based on your suggestion to use QVector2D, I reworked it to use QgsVector which does make for cleaner code. I actually was going to experiment with QgsVector earlier, but found that it wasn't available in Python Console and then I forgot about it. Also used your more descriptive name for the width variable. 

I did notice that you used rasterUnitsPerPixelX() twice instead of the X AND Y pixel units  in your comment on using QVector2D -- was that a test?

Thanks -- it does help to have more eyes on things.

**Comment by vmora** [comment]
> rasterUnitsPerPixelX() twice instead of the X AND Y pixel units in your comment on using QVector2D -- was that a test?

Nope, just a plain old mistake :) Sorry for that.

**Comment by slarosa** [comment]
@qgis/qgis-developers can we merge this?

**Comment by jef-n** [comment]
manually merge in 7f731ae

**Pull request closed without merging**